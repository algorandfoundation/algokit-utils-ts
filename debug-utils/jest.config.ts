import type { Config } from '@jest/types'

const config: Config.InitialOptions = {
  preset: 'ts-jest',
  testEnvironment: 'node',
  testMatch: ['**/*.spec.ts'],
  testTimeout: 20000,
  bail: 1,
  moduleDirectories: ['node_modules', 'src'],
  transform: {
    '^.+\\.ts?$': [
      'ts-jest',
      {
        tsconfig: 'tsconfig.test.json',
        diagnostics: {
          ignoreCodes: [1343],
        },
      },
    ],
  },
}
export default config
