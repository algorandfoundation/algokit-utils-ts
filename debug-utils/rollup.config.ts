// debug-utils/rollup.config.js
import baseConfig from '../rollup.config.js'

export default {
  ...baseConfig,
  input: 'src/index.ts',
  // Adjust other properties as needed
}
