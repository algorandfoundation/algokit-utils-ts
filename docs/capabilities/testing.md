# Automated testing

Automated testing is a higher-order use case capability provided by AlgoKit Utils that builds on top of the core capabilities. It allows you to use terse, robust automated testing primitives that work across any testing framework (including jest and vitest) to facilitate fixture management, quickly generating isolated and funded test accounts, transaction logging, indexer wait management and log capture.

To see some usage examples check out the all of the [automated tests](../../src/) and the various \*.spec.ts files (AlgoKit Utils [dogfoods](https://en.wikipedia.org/wiki/Eating_your_own_dog_food) it's own testing library). Alternatively, you can see an example of using this library to test a smart contract with [the tests](https://github.com/algorandfoundation/nft_voting_tool/blob/main/src/algorand/smart_contracts/tests/voting.spec.ts) for the [on-chain voting tool](https://github.com/algorandfoundation/nft_voting_tool#readme).

## Module import

The testing capability is not exposed from the [root algokit module](../code/modules/index.md) so there is a clear separation between testing functionality and non-testing functionality.

To access all of the functionality in the testing capability individually, you can import the [testing module](../code/modules/testing.md):

```typescript
import * as algotesting from '@algorandfoundation/algokit-utils/testing'
```

## Algorand fixture

In general, the only entrypoint you will need to use the testing capability is just by importing the `algorandFixture` since it exposes the rest of the functionality in a manner that is easy to integrate with an underlying test framework like Jest or vitest:

```typescript
import { algorandFixture } from '@algorandfoundation/algokit-utils/testing'
```

### Using with Jest

To integrate with [Jest](https://jestjs.io/) you need to pass the `fixture.beforeEach` method into Jest's `beforeEach` method and then within each test you can access `fixture.context` to access per-test isolated fixture values.

```typescript
import { describe, test, beforeEach } from '@jest/globals'
import { algorandFixture } from './testing'

describe('MY MODULE', () => {
  const fixture = algorandFixture()
  beforeEach(fixture.beforeEach, 10_000)

  test('MY TEST', async () => {
    const { algorand, testAccount /* ... */ } = fixture.context

    // Test stuff!
  })
})
```

Occasionally there may be a delay when first running the fixture setup so we add a 10s timeout to avoid intermittent test failures (`10_000`).

### Using with vitest

To integrate with [vitest](https://vitest.dev/) you need to pass the `fixture.beforeEach` method into vitest's `beforeEach` method and then within each test you can access `fixture.context` to access per-test isolated fixture values.

```typescript
import { describe, test, beforeEach } from 'vitest'
import { algorandFixture } from './testing'

describe('MY MODULE', () => {
  const fixture = algorandFixture()
  beforeEach(fixture.beforeEach, 10_000)

  test('MY TEST', async () => {
    const { algorand, testAccount /* ... */ } = fixture.context

    // Test stuff!
  })
})
```

### Fixture configuration

When calling `algorandFixture()` you can optionally pass in some fixture configuration, with any of these properties (all optional):

- `algod?: Algodv2` - An optional algod client, if not specified then it will create one against environment variables defined network (if present) or default LocalNet
- `indexer?: Indexer` - An optional indexer client, if not specified then it will create one against environment variables defined network (if present) or default LocalNet
- `kmd?: Kmd` - An optional kmd client, if not specified then it will create one against environment variables defined network (if present) or default LocalNet
- `testAccountFunding?: AlgoAmount` - The [amount](./amount.md) of funds to allocate to the default testing account, if not specified then it will get `10` Algos

### Using the fixture context

The `fixture.context` property is of type [`AlgorandTestAutomationContext`](../code/interfaces/types_testing.AlgorandTestAutomationContext.md) exposes the following properties from which you can pick which ones you want in a given test using an object [destructuring assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment):

- `algorand: AlgorandClient` - An [`AlgorandClient`](./algorand-client.md) instance
- `algod: Algodv2` - Proxy Algod client instance that will log sent transactions in `transactionLogger`
- `indexer: Indexer` - Indexer client instance
- `kmd: Kmd` - KMD client instance
- `transactionLogger: TransactionLogger` - Transaction logger that will log transaction IDs for all transactions issued by `algod`
- `testAccount: Account` - Funded test account that is ephemerally created for each test
- `generateAccount: (params: GetTestAccountParams) => Promise<Account>` - Generate and fund an additional ephemerally created account
- `waitForIndexer: () => Promise<void>` - Wait for the indexer to catch up with all transactions logged by transactionLogger
- `waitForIndexerTransaction: (transactionId: string) => Promise<TransactionLookupResult>` - Wait for the indexer to catch up with the given transaction ID

## Log capture fixture

If you want to capture log messages from AlgoKit that are issued within your test so that you can assert on them or parse them for debugging information etc. then you can use the log capture fixture.

```typescript
import { algoKitLogCaptureFixture } from '@algorandfoundation/algokit-utils/testing'
```

The log capture fixture works by setting the logger within the AlgoKit configuration to be a [`TestLogger`](../code/classes/testing.TestLogger.md) during the test run.

### Using with Jest

To integrate with [Jest](https://jestjs.io/) you need to pass the `fixture.beforeEach` method into Jest's `beforeEach` method and then within each test you can access `fixture.context` to access per-test isolated fixture values.

```typescript
import { describe, test, beforeEach, afterEach } from '@jest/globals'
import { algoKitLogCaptureFixture } from './testing'

describe('MY MODULE', () => {
  const logs = algoKitLogCaptureFixture()
  beforeEach(logs.beforeEach)
  afterEach(logs.afterEach)

  test('MY TEST', async () => {
    const { algorand, testAccount } = fixture.context
    // Test stuff!

    const capturedLogs = logs.testLogger.capturedLogs
    // do stuff with the logs
  })
})
```

### Using with vitest

To integrate with [vitest](https://vitest.dev/) you need to pass the `fixture.beforeEach` method into vitest's `beforeEach` method and then within each test you can access `fixture.context` to access per-test isolated fixture values.

```typescript
import { describe, test, beforeEach, afterEach } from 'vitest'
import { algoKitLogCaptureFixture } from './testing'

describe('MY MODULE', () => {
  const logs = algoKitLogCaptureFixture()
  beforeEach(logs.beforeEach)
  afterEach(logs.afterEach)

  test('MY TEST', async () => {
    const { algorand, testAccount } = fixture.context
    // Test stuff!

    const capturedLogs = logs.testLogger.capturedLogs
    // do stuff with the logs
  })
})
```

### Snapshot testing the logs

If you want to quickly pin some behaviour of what logic you have does in terms of invoking AlgoKit methods you can do a [snapshot test](https://jestjs.io/docs/snapshot-testing) / [approval test](https://approvaltests.com/) of the captured log output. The only problem is this output will contain identifiers that will change for every test run and thus will constantly break the snapshot. In order to work around this you can use the `getLogSnapshot` method on the `TestLogger`, which will replace those changing values with predictable strings to keep the snapshot integrity intact.

This might look something like this:

```typescript
const { algorand, testAccount } = fixture.context
const result = await algorand.client.getTypedClientById(HelloWorldContractClient, { id: 0 }).deploy()
expect(
  logging.testLogger.getLogSnapshot({
    accounts: [testAccount],
    transactions: [result.transaction],
    apps: [result.appId],
  }),
).toMatchSnapshot()
```

## Waiting for indexer

Often there will be things that you do in your test that you may want to assert in using data that is exclusively in indexer such as transaction notes. The problem is indexer asynchronously indexes the data in algod, even when devmode is turned on and algod instantly confirms transactions.

This means it's easy to create tests that are flaky and have intermittent test failures (sometimes indexer is up to date and other times it hasn't caught up yet).

The testing capability provides mechanisms for waiting for indexer to catch up, namely:

- `algotesting.runWhenIndexerCaughtUp(run: () => Promise<T>)` - Executes the given action every 200ms up to 20 times until there is no longer an error with a `status` property with `404` and then returns the result of the action; this will work for any call that calls indexer APIs expecting to return a single record
- `algorandFixture.waitForIndexer()` - Waits for indexer to catch up with all transactions that have been captured by the `transactionLogger` in the Algorand fixture
- `algorandFixture.waitForIndexerTransaction(transactionId)` - Waits for indexer to catch up with the single transaction of the given ID

## Logging transactions

When testing, it can be useful to capture all of the transactions that have been issued with a given test run. They can then be asserted on, or used for [waiting for indexer](#waiting-for-indexer), etc.

The testing capability provides the ability to capture transactions via the [`TransactionLogger`](../code/classes/testing.TransactionLogger.md) class.

The `TransactionLogger` has the following methods:

- `logRawTransaction(signedTransactions: Uint8Array | Uint8Array[])` - Logs the IDs of the given signed transaction(s)
- `capture(algod)` - Returns a proxy `algosdk.Algodv2` instance that wraps the given `algod` client that will call `logRawTransaction` for every call to `sendRawTransaction` on that algod instance
- `sentTransactionIds` - Returns the currently captured list of transaction IDs that have been logged
- `clear()` - Clears the current list of transaction IDs
- `waitForIndexer(indexer)` - [Waits for the given indexer instance to catch up](#waiting-for-indexer) with the currently logged transaction IDs

The easiest way to use this functionality is via the [Algorand fixture](#algorand-fixture), which automatically provides a `transactionLogger` and a proxy `algod` connected to that `transactionLogger`.

## Getting a test account

When testing, it's often useful to ephemerally generate random accounts, fund them with some number of Algos and then use that account to perform transactions. By creating an ephemeral, random account you naturally get isolation between tests and test runs and don't need to start from a specific blockchain network state. This makes test less flakey, and also means the same test can be run against LocalNet and (say) TestNet.

The key when generating a test account is getting hold of a [dispenser](./transfer.md#dispenser) and then [ensuring the test account is funded](./transfer.md#ensurefunded).

To make it easier to quickly get a test account the testing capability provides the following mechanisms:

- [`algotesting.getTestAccount(testAccountParams, algod, kmd?)`](../code/modules/testing.md#gettestaccount) - Generates a random new account, logs the mnemonic of the account (unless suppressed), funds it from the [dispenser](./transfer.md#dispenser)
- `algorandFixture.testAccount` - A test account that is always generated for every test (log output suppressed to reduce noise, but worth noting that means the mnemonic isn't logged for this account), by default it is given 10 Algos unless overridden in the [fixture config](#fixture-configuration)
- [`algorandFixture.generateAccount(testAccountParams)`](../code/interfaces/types_testing.AlgorandTestAutomationContext.md#generateaccount) - Allows you to quickly generate a test account with the `algod` and `kmd` instances that are part of the given fixture

The parameters object that controls test account generation, [`GetTestAccountParams`](../code/interfaces/types_testing.GetTestAccountParams.md), has the following properties:

- `initialFunds: AlgoAmount` - Initial funds to ensure the account has
- `suppressLog?: boolean` - Whether to suppress the log (which includes a mnemonic) or not (default: do not suppress the log)
