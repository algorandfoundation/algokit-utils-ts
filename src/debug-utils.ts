import algosdk from 'algosdk'
import * as fs from 'fs'
import * as path from 'path'
import { Config } from '.'
import { performAtomicTransactionComposerSimulate } from './transaction'
import {
  AVMDebuggerSourceMap,
  DEBUG_TRACES_DIR,
  PersistSourceMapInput,
  TRACES_FILE_EXT,
  buildAVMSourcemap,
  isNode,
  upsertDebugSourcemaps,
} from './types/debug-utils'

/**
 * This function persists the source maps for the given sources.
 *
 * @param sources An array of objects, each containing the teal content, app name, and file name.
 * @param projectRoot The root directory of the project.
 * @param client An Algod client to perform the compilation.
 * @param withSources A boolean indicating whether to include the source files in the output.
 *
 * @returns A promise that resolves when the source maps have been persisted.
 */
async function persistSourcemaps({
  sources,
  projectRoot,
  client,
  withSources,
}: {
  sources: PersistSourceMapInput[]
  projectRoot: string
  client: algosdk.Algodv2
  withSources?: boolean
}): Promise<void> {
  if (!isNode()) {
    throw new Error('Sourcemaps can only be persisted in Node.js environment.')
  }

  try {
    const sourceMaps = await Promise.all(
      sources.map((source) =>
        buildAVMSourcemap({
          tealContent: source.teal,
          appName: source.appName,
          fileName: source.fileName,
          outputPath: projectRoot,
          client: client,
          withSources: withSources,
        }),
      ),
    )

    await upsertDebugSourcemaps(sourceMaps, projectRoot)
  } catch (error) {
    const err = error instanceof Error ? error : new Error(String(error))
    Config.getLogger().error(`Failed to persist avm sourceMaps: ${err.stack ?? err.message ?? err}.`)
  }
}

/**
 * This function simulates the atomic transactions using the provided `AtomicTransactionComposer` object and `Algodv2` object,
 * and persists the simulation response to an AVM Debugger compliant JSON file.
 *
 * @param atc The AtomicTransactionComposer with transaction(s) loaded.
 * @param algod An Algod client to perform the simulation.
 * @param projectRoot The root directory of the project.
 * @param bufferSizeMb The buffer size in megabytes.
 *
 * @returns The simulation result, which includes various details about how the transactions would be processed.
 *
 * @example
 * const atc = new AtomicTransactionComposer();
 * const algod = new algosdk.Algodv2(token, server, port);
 * const projectRoot = '/path/to/project';
 * const bufferSizeMb = 10;
 *
 * const result = await simulateAndPersistResponse({ atc, projectRoot, algod, bufferSizeMb });
 * console.log(result);
 */
async function simulateAndPersistResponse({
  atc,
  projectRoot,
  algod,
  bufferSizeMb,
}: {
  atc: algosdk.AtomicTransactionComposer
  projectRoot: string
  algod: algosdk.Algodv2
  bufferSizeMb: number
}) {
  if (!isNode()) {
    throw new Error('Sourcemaps can only be persisted in Node.js environment.')
  }

  try {
    const simulateResult = await performAtomicTransactionComposerSimulate(atc, algod)
    const txnGroups = simulateResult.txnGroups

    const txnTypesCount = txnGroups.reduce((acc: Record<string, number>, txnGroup) => {
      const txnType = txnGroup.txnResults[0].txnResult.txn.txn.type
      acc[txnType] = (acc[txnType] || 0) + 1
      return acc
    }, {})

    const txnTypesStr = Object.entries(txnTypesCount)
      .map(([type, count]) => `${count}#${type}`)
      .join('_')

    const timestamp = new Date().toISOString().replace(/[:.]/g, '')
    const outputRootDir = path.join(projectRoot, DEBUG_TRACES_DIR)
    const outputFileName = `${timestamp}_lr${simulateResult.lastRound}_${txnTypesStr}${TRACES_FILE_EXT}`
    const outputFilePath = path.join(outputRootDir, outputFileName)

    if (!fs.existsSync(path.dirname(outputFilePath))) {
      await fs.promises.mkdir(path.dirname(outputFilePath), { recursive: true })
    }

    // cleanup old files if buffer size is exceeded
    let totalSize = (
      await Promise.all(
        (await fs.promises.readdir(outputRootDir)).map(async (file) => (await fs.promises.stat(path.join(outputRootDir, file))).size),
      )
    ).reduce((a, b) => a + b, 0)

    if (totalSize > bufferSizeMb * 1024 * 1024) {
      const sortedFiles = (await fs.promises.readdir(outputRootDir)).map(async (file) => {
        const stats = await fs.promises.stat(path.join(outputRootDir, file))
        return { file, mtime: stats.mtime }
      })

      // Since map returns an array of promises, we need to await all of them
      const resolvedFiles = await Promise.all(sortedFiles)

      resolvedFiles.sort((a, b) => a.mtime.getTime() - b.mtime.getTime())

      while (totalSize > bufferSizeMb * 1024 * 1024) {
        const oldestFilePromise = sortedFiles.shift()
        if (oldestFilePromise) {
          const oldestFile = await oldestFilePromise
          const stats = await fs.promises.stat(path.join(outputRootDir, oldestFile.file))
          totalSize -= stats.size
          await fs.promises.unlink(path.join(outputRootDir, oldestFile.file))
        }
      }
    }

    await fs.promises.writeFile(outputFilePath, JSON.stringify(simulateResult, null, 2))
  } catch (error) {
    const err = error instanceof Error ? error : new Error(String(error))
    Config.getLogger().error(`Failed to simulate and persist avm traces: ${err.stack ?? err.message ?? err}.`)
  }
}

export { AVMDebuggerSourceMap, PersistSourceMapInput, persistSourcemaps, simulateAndPersistResponse }
