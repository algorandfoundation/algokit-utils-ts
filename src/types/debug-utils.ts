import algosdk from 'algosdk'
import * as crypto from 'crypto'
import * as fs from 'fs'
import * as path from 'path'

export const ALGOKIT_DIR = '.algokit'
export const SOURCES_DIR = 'sources'
export const SOURCES_FILE = 'sources.avm.json'
export const TRACES_FILE_EXT = '.trace.avm.json'
export const DEBUG_TRACES_DIR = 'debug_traces'
export const TEAL_FILE_EXT = '.teal'
export const TEAL_SOURCEMAP_EXT = '.teal.tok.map'

export interface ErrnoException extends Error {
  errno?: number
  code?: string
  path?: string
  syscall?: string
}

export interface AVMDebuggerSourceMapDict {
  'txn-group-sources': Array<{
    'sourcemap-location': string
    hash: string
  }>
}

export class AVMDebuggerSourceMapEntry {
  constructor(
    public location: string,
    public programHash: string,
  ) {}

  equals(other: AVMDebuggerSourceMapEntry): boolean {
    return this.location === other.location && this.programHash === other.programHash
  }

  toString(): string {
    return JSON.stringify({ 'sourcemap-location': this.location, hash: this.programHash })
  }
}

export class AVMDebuggerSourceMap {
  txnGroupSources: AVMDebuggerSourceMapEntry[]

  constructor(txnGroupSources: AVMDebuggerSourceMapEntry[]) {
    this.txnGroupSources = txnGroupSources
  }

  static fromDict(data: AVMDebuggerSourceMapDict): AVMDebuggerSourceMap {
    return new AVMDebuggerSourceMap(
      data['txn-group-sources'].map((item) => new AVMDebuggerSourceMapEntry(item['sourcemap-location'], item['hash'])),
    )
  }

  toDict(): AVMDebuggerSourceMapDict {
    return { 'txn-group-sources': this.txnGroupSources.map((item) => JSON.parse(item.toString())) }
  }
}

export class PersistSourceMapInput {
  teal: string
  appName: string
  private _fileName: string

  constructor(teal: string, appName: string, fileName: string) {
    this.teal = teal
    this.appName = appName
    this._fileName = this.stripTealExtension(fileName)
  }

  get fileName(): string {
    return this._fileName
  }

  set fileName(value: string) {
    this._fileName = this.stripTealExtension(value)
  }

  private stripTealExtension(fileName: string): string {
    if (fileName.endsWith('.teal')) {
      return fileName.slice(0, -5)
    }
    return fileName
  }
}

export async function loadOrCreateSources(sourcesPath: string): Promise<AVMDebuggerSourceMap> {
  try {
    const data = JSON.parse(await fs.promises.readFile(sourcesPath, 'utf8'))
    return AVMDebuggerSourceMap.fromDict(data)
  } catch (error: unknown) {
    const err = error as ErrnoException

    if (err.code === 'ENOENT') {
      return new AVMDebuggerSourceMap([])
    } else {
      throw error
    }
  }
}

export async function upsertDebugSourcemaps(sourceMaps: AVMDebuggerSourceMapEntry[], projectRoot: string): Promise<void> {
  const sourcesPath = path.join(projectRoot, ALGOKIT_DIR, SOURCES_DIR, SOURCES_FILE)
  const sources = await loadOrCreateSources(sourcesPath)

  for (const sourcemap of sourceMaps) {
    const sourceFilePath = path.resolve(sourcemap.location)
    try {
      await fs.promises.access(sourceFilePath)
      const index = sources.txnGroupSources.findIndex((item) => item.equals(sourcemap))
      if (index === -1) {
        sources.txnGroupSources.push(sourcemap)
      } else {
        sources.txnGroupSources[index] = sourcemap
      }
    } catch (error: unknown) {
      const err = error as ErrnoException

      if (err.code === 'ENOENT') {
        const index = sources.txnGroupSources.findIndex((item) => item.equals(sourcemap))
        if (index !== -1) {
          sources.txnGroupSources.splice(index, 1)
        }
      } else {
        throw error
      }
    }
  }

  await fs.promises.writeFile(sourcesPath, JSON.stringify(sources.toDict()), 'utf8')
}

export async function writeToFile(filePath: string, content: string): Promise<void> {
  await fs.promises.mkdir(path.dirname(filePath), { recursive: true })
  await fs.promises.writeFile(filePath, content, 'utf8')
}

export async function buildAVMSourcemap({
  tealContent,
  appName,
  fileName,
  outputPath,
  client,
  withSources = true,
}: {
  tealContent: string
  appName: string
  fileName: string
  outputPath: string
  client: algosdk.Algodv2
  withSources?: boolean
}): Promise<AVMDebuggerSourceMapEntry> {
  const result = await client.compile(tealContent).sourcemap(true).do()
  const programHash = crypto.createHash('SHA-512/256').update(Buffer.from(result.result, 'base64')).digest('base64')
  const sourceMap = result.sourcemap
  sourceMap.sources = withSources ? [`${fileName}${TEAL_FILE_EXT}`] : []

  const outputDirPath = path.join(outputPath, ALGOKIT_DIR, SOURCES_DIR, appName)
  const sourceMapOutputPath = path.join(outputDirPath, `${fileName}${TEAL_SOURCEMAP_EXT}`)
  const tealOutputPath = path.join(outputDirPath, `${fileName}${TEAL_FILE_EXT}`)
  await writeToFile(sourceMapOutputPath, JSON.stringify(sourceMap))

  if (withSources) {
    await writeToFile(tealOutputPath, tealContent)
  }

  return new AVMDebuggerSourceMapEntry(sourceMapOutputPath, programHash)
}

// simple function checking whether this is running in node or browser environment
export function isNode(): boolean {
  return typeof window === 'undefined'
}
