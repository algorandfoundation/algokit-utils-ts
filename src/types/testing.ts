import algosdk from 'algosdk'
import { TransactionLogger } from '../testing'
import { TestLogger } from '../testing/test-logger'
import { AlgoAmount } from '../types/amount'
import { SendTransactionFrom } from '../types/transaction'
import { TransactionSignerAccount } from './account'
import AlgorandClient from './algorand-client'
import { TransactionLookupResult } from './indexer'
import { AlgoConfig } from './network-client'
import Account = algosdk.Account
import Algodv2 = algosdk.Algodv2
import Indexer = algosdk.Indexer
import Kmd = algosdk.Kmd
import Transaction = algosdk.Transaction

/**
 * Test automation context.
 */
export interface AlgorandTestAutomationContext {
  /** An AlgorandClient instance loaded with the current context, including testAccount and any generated accounts loaded as signers */
  algorand: AlgorandClient
  /** Algod client instance that will log transactions in `transactionLogger` */
  algod: Algodv2
  /** Indexer client instance */
  indexer: Indexer
  /** KMD client instance */
  kmd: Kmd
  /** Transaction logger that will log transaction IDs for all transactions issued by `algod` */
  transactionLogger: TransactionLogger
  /** Default, funded test account that is ephemerally created */
  testAccount: Account & TransactionSignerAccount
  /** Generate and fund an additional ephemerally created account */
  generateAccount: (params: GetTestAccountParams) => Promise<Account & TransactionSignerAccount>
  /** Wait for the indexer to catch up with all transactions logged by `transactionLogger` */
  waitForIndexer: () => Promise<void>
  /** Wait for the indexer to catch up with the given transaction ID */
  waitForIndexerTransaction: (transactionId: string) => Promise<TransactionLookupResult>
}

/**
 * Parameters for the `getTestAccount` function.
 */
export interface GetTestAccountParams {
  /** Initial funds to ensure the account has */
  initialFunds: AlgoAmount
  /** Whether to suppress the log (which includes a mnemonic) or not (default: do not suppress the log) */
  suppressLog?: boolean
  /** Optional override for how to get a test account; this allows you to retrieve accounts from a known or cached list of accounts. */
  accountGetter?: (algorand: AlgorandClient) => Promise<Account>
}

/** Configuration for creating an Algorand testing fixture. */
export interface AlgorandFixtureConfig extends Partial<AlgoConfig> {
  /** An optional algod client, if not specified then it will create one against `algodConfig` (if present) then environment variables defined network (if present) or default LocalNet. */
  algod?: Algodv2
  /** An optional indexer client, if not specified then it will create one against `indexerConfig` (if present) then environment variables defined network (if present) or default LocalNet. */
  indexer?: Indexer
  /** An optional kmd client, if not specified then it will create one against `kmdConfig` (if present) then environment variables defined network (if present) or default LocalNet. */
  kmd?: Kmd
  /** The amount of funds to allocate to the default testing account, if not specified then it will get 10 Algos. */
  testAccountFunding?: AlgoAmount
  /** Optional override for how to get an account; this allows you to retrieve accounts from a known or cached list of accounts. */
  accountGetter?: (algod: Algodv2, kmd?: Kmd) => Promise<Account>
}

/** An Algorand automated testing fixture */
export interface AlgorandFixture {
  /**
   * Retrieve the current context.
   * Useful with destructuring.
   * @example
   * ```typescript
   * test('My test', () => {
   *     const {algod, indexer, testAccount, ...} = algorand.context
   * })
   * ```
   */
  get context(): AlgorandTestAutomationContext

  /**
   * Retrieve an `AlgorandClient` loaded with the current context, including testAccount and any generated accounts loaded as signers.
   */
  get algorand(): AlgorandClient

  /**
   * Testing framework agnostic handler method to run before each test to prepare the `context` for that test.
   */
  beforeEach: () => Promise<void>
}

/** Configuration for preparing a captured log snapshot.
 * This helps ensure that the provided configuration items won't appear
 *  with random values in the log snapshot, but rather will get substituted with predictable ids.
 */
export interface LogSnapshotConfig {
  /** Any transaction IDs or transactions to replace the ID for predictably */
  transactions?: (string | Transaction)[]
  /** Any accounts/addresses to replace the address for predictably */
  accounts?: (string | SendTransactionFrom)[]
  /** Any app IDs to replace predictably */
  apps?: (string | number | bigint)[]
}

export interface AlgoKitLogCaptureFixture {
  /** The test logger instance for the current test */
  get testLogger(): TestLogger
  /**
   * Testing framework agnostic handler method to run before each test to prepare the `testLogger` for that test.
   */
  beforeEach: () => void
  /**
   * Testing framework agnostic handler method to run after each test to reset the logger.
   */
  afterEach: () => void
}
