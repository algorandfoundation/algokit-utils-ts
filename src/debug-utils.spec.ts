import { describe, test } from '@jest/globals'
import algosdk, { AtomicTransactionComposer } from 'algosdk'
import * as fsSync from 'fs'
import * as fs from 'fs/promises'
import * as os from 'os'
import * as path from 'path'
import { AVMDebuggerSourceMap, PersistSourceMapInput, persistSourceMaps, simulateAndPersistResponse } from './debug-utils'
import { algorandFixture } from './testing'

async function fileExists(filePath: string): Promise<boolean> {
  try {
    await fs.access(filePath)
    return true
  } catch {
    return false
  }
}

describe('debug utils tests', () => {
  const localnet = algorandFixture()
  const timeout = 10e6

  beforeAll(localnet.beforeEach)

  test(
    'build teal sourceMaps',
    async () => {
      const { algod } = localnet.context
      const cwd = await fs.mkdtemp(path.join(os.tmpdir(), 'cwd'))

      const approval = `
#pragma version 9
int 1
`
      const clear = `
#pragma version 9
int 1
`
      const sources = [
        new PersistSourceMapInput(approval, 'cool_app', 'approval.teal'),
        new PersistSourceMapInput(clear, 'cool_app', 'clear'),
      ]

      await persistSourceMaps({ sources: sources, projectRoot: cwd, client: algod })

      const rootPath = path.join(cwd, '.algokit', 'sources')
      const sourcemapFilePath = path.join(rootPath, 'sources.avm.json')
      const appOutputPath = path.join(rootPath, 'cool_app')

      expect(await fileExists(sourcemapFilePath)).toBeTruthy()
      expect(await fileExists(path.join(appOutputPath, 'approval.teal'))).toBeTruthy()
      expect(await fileExists(path.join(appOutputPath, 'approval.teal.tok.map'))).toBeTruthy()
      expect(await fileExists(path.join(appOutputPath, 'clear.teal'))).toBeTruthy()
      expect(await fileExists(path.join(appOutputPath, 'clear.teal.tok.map'))).toBeTruthy()

      // Remove location from sourcemap to make snapshot deterministic
      const result = AVMDebuggerSourceMap.fromDict(JSON.parse(await fs.readFile(sourcemapFilePath, 'utf8')))
      for (const item of result.txnGroupSources) {
        item.location = 'dummy'
      }

      expect(result).toMatchSnapshot()

      // check for updates in case of multiple runs
      await persistSourceMaps({ sources: sources, projectRoot: cwd, client: algod })
      const resultAfterUpdate = AVMDebuggerSourceMap.fromDict(JSON.parse(await fs.readFile(sourcemapFilePath, 'utf8')))
      for (const item of resultAfterUpdate.txnGroupSources) {
        expect(item.location).not.toBe('dummy')
      }
    },
    timeout,
  )

  test(
    'build teal sourceMaps without sources',
    async () => {
      const { algod } = localnet.context
      const cwd = await fs.mkdtemp(path.join(os.tmpdir(), 'cwd'))

      const approval = `
#pragma version 9
int 1
`
      const clear = `
#pragma version 9
int 1
`
      const sources = [
        new PersistSourceMapInput(approval, 'cool_app', 'approval.teal'),
        new PersistSourceMapInput(clear, 'cool_app', 'clear'),
      ]

      await persistSourceMaps({ sources: sources, projectRoot: cwd, client: algod, withSources: false })

      const rootPath = path.join(cwd, '.algokit', 'sources')
      const sourcemapFilePath = path.join(rootPath, 'sources.avm.json')
      const appOutputPath = path.join(rootPath, 'cool_app')

      expect(await fileExists(sourcemapFilePath)).toBeTruthy()
      expect(await fileExists(path.join(appOutputPath, 'approval.teal'))).toBeFalsy()
      expect(await fileExists(path.join(appOutputPath, 'approval.teal.tok.map'))).toBeTruthy()
      expect(JSON.parse(await fs.readFile(path.join(appOutputPath, 'approval.teal.tok.map'), 'utf8')).sources).toEqual([])
      expect(await fileExists(path.join(appOutputPath, 'clear.teal'))).toBeFalsy()
      expect(await fileExists(path.join(appOutputPath, 'clear.teal.tok.map'))).toBeTruthy()
      expect(JSON.parse(await fs.readFile(path.join(appOutputPath, 'clear.teal.tok.map'), 'utf8')).sources).toEqual([])

      const result = AVMDebuggerSourceMap.fromDict(JSON.parse(await fs.readFile(sourcemapFilePath, 'utf8')))
      for (const item of result.txnGroupSources) {
        item.location = 'dummy'
      }

      expect(result).toMatchSnapshot()
    },
    timeout,
  )

  test(
    'simulate and persist response',
    async () => {
      const { algod, testAccount } = localnet.context
      const cwd = await fs.mkdtemp(path.join(os.tmpdir(), 'cwd'))

      const transaction = algosdk.makePaymentTxnWithSuggestedParamsFromObject({
        from: testAccount.addr,
        to: testAccount.addr,
        rekeyTo: undefined,
        amount: 0,
        suggestedParams: await localnet.context.algod.getTransactionParams().do(),
        closeRemainderTo: undefined,
      })
      const atc = new AtomicTransactionComposer()
      atc.addTransaction({ txn: transaction, signer: algosdk.makeBasicAccountTransactionSigner(testAccount) })
      const bufferSizeMb = 256

      await simulateAndPersistResponse({ atc: atc, projectRoot: cwd, algod: algod, bufferSizeMb: bufferSizeMb })

      const debugTracesPath = path.join(cwd, 'debug_traces')
      const files = await fs.readdir(debugTracesPath)

      const filePath = path.join(debugTracesPath, files[0])
      const content = JSON.parse(await fs.readFile(filePath, 'utf8'))
      expect(content.txnGroups[0].txnResults[0].txnResult.txn.txn.type).toBe('pay')
      expect(content).toHaveProperty('lastRound')

      while (fsSync.existsSync(filePath)) {
        const tmpAtc = atc.clone()
        await simulateAndPersistResponse({ atc: tmpAtc, projectRoot: cwd, algod: algod, bufferSizeMb: 0.01 })
      }
    },
    timeout,
  )
})
