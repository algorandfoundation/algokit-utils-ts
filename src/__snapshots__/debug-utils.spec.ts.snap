// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debug utils tests build teal sourceMaps 1`] = `
AVMDebuggerSourceMap {
  "txnGroupSources": [
    AVMDebuggerSourceMapEntry {
      "location": "dummy",
      "programHash": "EC1P8unO+zjVbdF8XZOs1rp+uaGNk7vXtZ/IYsN/sug=",
    },
    AVMDebuggerSourceMapEntry {
      "location": "dummy",
      "programHash": "EC1P8unO+zjVbdF8XZOs1rp+uaGNk7vXtZ/IYsN/sug=",
    },
  ],
}
`;

exports[`debug utils tests build teal sourceMaps without sources 1`] = `
AVMDebuggerSourceMap {
  "txnGroupSources": [
    AVMDebuggerSourceMapEntry {
      "location": "dummy",
      "programHash": "EC1P8unO+zjVbdF8XZOs1rp+uaGNk7vXtZ/IYsN/sug=",
    },
    AVMDebuggerSourceMapEntry {
      "location": "dummy",
      "programHash": "EC1P8unO+zjVbdF8XZOs1rp+uaGNk7vXtZ/IYsN/sug=",
    },
  ],
}
`;
